pkg/resourcemonitor/podresourcesscanner.go

```diff
@@ -0,0 +1,191 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resourcemonitor
+
+import (
+	"context"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog/v2"
+	podresourcesapi "k8s.io/kubelet/pkg/apis/podresources/v1"
+	"sigs.k8s.io/node-feature-discovery/pkg/apihelper"
+)
+
+type PodResourcesScanner struct {
+	namespace         string
+	podResourceClient podresourcesapi.PodResourcesListerClient
+	apihelper         apihelper.APIHelpers
+}
+
+func NewPodResourcesScanner(namespace string, podResourceClient podresourcesapi.PodResourcesListerClient, kubeApihelper apihelper.APIHelpers) (ResourcesScanner, error) {
+	resourcemonitorInstance := &PodResourcesScanner{
+		namespace:         namespace,
+		podResourceClient: podResourceClient,
+		apihelper:         kubeApihelper,
+	}
+	if resourcemonitorInstance.namespace != "" {
+		klog.Infof("watching namespace %q", resourcemonitorInstance.namespace)
+	} else {
+		klog.Infof("watching all namespaces")
+	}
+
+	return resourcemonitorInstance, nil
+}
+
+// isWatchable tells if the the given namespace should be watched.
+func (resMon *PodResourcesScanner) isWatchable(podNamespace string, podName string, hasDevice bool) (bool, bool, error) {
+	cli, err := resMon.apihelper.GetClient()
+	if err != nil {
+		return false, false, err
+	}
+	pod, err := resMon.apihelper.GetPod(cli, podNamespace, podName)
+	if err != nil {
+		return false, false, err
+	}
+
+	klog.Infof("podresource: %s", podName)
+	isIntegralGuaranteed := podGuaranteedCPUs(pod)
+
+	if resMon.namespace == "" && (isIntegralGuaranteed || hasDevice) {
+		return true, isIntegralGuaranteed, nil
+	}
+	// TODO:  add an explicit check for guaranteed pods and pods with devices
+	return resMon.namespace == podNamespace && (isIntegralGuaranteed || hasDevice), isIntegralGuaranteed, nil
+}
+
+func podGuaranteedCPUs(pod *v1.Pod) bool {
+	var totalCPU int64
+	var cpuQuantity resource.Quantity
+	for _, container := range pod.Spec.InitContainers {
+
+		var ok bool
+		if cpuQuantity, ok = container.Resources.Requests[v1.ResourceCPU]; !ok {
+			continue
+		}
+		totalCPU += cpuQuantity.Value()
+		isInitContainerGuaranteed := guaranteedCPUs(pod, &container)
+		if !isInitContainerGuaranteed {
+			return false
+		}
+	}
+	for _, container := range pod.Spec.Containers {
+		var ok bool
+		if cpuQuantity, ok = container.Resources.Requests[v1.ResourceCPU]; !ok {
+			continue
+		}
+		totalCPU += cpuQuantity.Value()
+		isAppContainerGuaranteed := guaranteedCPUs(pod, &container)
+		if !isAppContainerGuaranteed {
+			return false
+		}
+	}
+
+	//No CPUs requested in all the containers in the pod
+	return totalCPU != 0
+}
+
+func guaranteedCPUs(pod *v1.Pod, container *v1.Container) bool {
+
+	cpuQuantity := container.Resources.Requests[v1.ResourceCPU]
+	return cpuQuantity.Value()*1000 == cpuQuantity.MilliValue()
```

Comment:
What does "guaranteed" stand for here? Shouldn't we inspect limits, too(?)

Reply:
Here the term “Guaranteed” is in the context where we check if a guaranteed pod is requesting integral CPUs basically to check if the CPUs are exclusively allocated in order to track the resources of that pod.
We already check for the pod QoS isWatchable() function and proceed if the pod is guaranteed so if the control reaches here it means that request is equal to limit and hence we only check for request.